### DISCUSSION TITLE
How can i add adsense ads units in nextjs project

### DISCUSSION THREAD

**Discussion by gsahab0007** [vercel/next.js#57181] [Help] [answered]
### Summary

I have already added google adsense auto ads code successfully, but i want to add adsense ads units in website pages made in next js latest version. Kindly help me in this regard.
Thanks in advance.

**Comment by shriekdj** [comment, marked as answer]
For App Router

```tsx
'use client';
import React, { Fragment, ReactNode, useEffect } from 'react';
import { usePathname, useSearchParams } from 'next/navigation';

type Props = {
  children: ReactNode;
};

declare global {
  interface Window {
    adsbygoogle?: any | any[];
  }
}

const GoogleAdUnit = ({ children }: Props) => {
  const pathname = usePathname();
  const searchParams = useSearchParams();
  useEffect(() => {
    try {
      (window.adsbygoogle = window.adsbygoogle || []).push({});
    } catch (err) {
      console.error(err);
    }
  }, [pathname, searchParams]);
  return <Fragment>{children}</Fragment>;
}


export default GoogleAdUnit;
```

And Use That Component with children as ins script

**Reply by batchor** [reply]
> And Use That Component with children as ins script

What do you mean by "ins script"?

**Reply by shriekdj** [reply]
@batchor When try to get google addsense ad unit code it gave the code which is `<ins>` html tags. like this 
```html
<ins class="adsbygoogle"
     style="display:block;"
     data-ad-client="ca-pub-1234"
     data-ad-host="ca-host-pub-5678"
     data-ad-format="auto">
</ins>
```


You Just have to Change that to like below
```jsx
<GoogleAdUnit>
<ins
  className="adsbygoogle"
  style={{ display: "block" }}
  data-ad-client="ca-pub-1234"
  data-ad-host="ca-host-pub-5678"
  data-ad-format="auto"
></ins>
</GoogleAdUnit>
```

**Comment by shriekdj** [comment]
if you are using pages router use `next-google-adsense` package instead

**Comment by gsahab0007** [comment]
Thanks bro .............great.......

**Comment by gsahab0007** [comment]
Any blog for full placement of code for ad units of adsense in next js latest version plz.....

**Reply by gsahab0007** [reply]
@shriekdj  bro thanks again........I implemented in my website.............

**Comment by shriekdj** [comment]
> @shriekdj bro thanks again........I implemented in my website.............

No Problem @gsahab0007 , i also had that issue :).

**Comment by calcgenie** [comment]
i never tired this one, previous i have a tried various set of codes but didnt help for this site https://www.calcgenie.com/distance-calculators/cm-to-inch. this one is my last hope. because perfromance of my site is hurting because of this

**Reply by shriekdj** [reply]
@calcgenie  I Would Suggest Use 2 Files Here for Performance Improvement, One Is Client Side Component Like `GoogleAdUnitClient` above code and Create Other Server Side Component `GoogleAdUnit` Like Below.
This Also Includes Code to Run Only in Production

```typescript
import React, { Fragment, ReactNode, Suspense } from 'react';
import GoogleAdUnitClient from './GoogleAdUnitClient';

type Props = {
  children: ReactNode;
};

const isProduction = env.NODE_ENV === 'production';

const GoogleAdUnit = ({ children }: Props) => {
  return (
    <Suspense>
      {
        isProduction ?
        // only run on Live Website
        <GoogleAdUnitClient>
          {children}
        </GoogleAdUnitClient> :

        // don't need to load unit code
        <Fragment>
          {children}
        </Fragment>
      }
    </Suspense>
  );
};

export default GoogleAdUnit;
```

**Reply by shriekdj** [reply]
@calcgenie  All The Other Packages Use `useRouter` from `next/router` which only works in pages router, but NextJS 13 Uses App Router Which Uses `useRouter` from `next/navigation`.

This is reason why old next.js packages for google adsense ad unit are not working, I think.
